### PR TITLE
Fix prod double backward when there are 2+ zeros

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113969

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### CODE REVIEW

**Review by albanD — CHANGES_REQUESTED** [review]
Any details?
Also this is most likely going to be a challenging change just for perf reasons.

**Comment by guilhermeleobas** [comment]
@albanD, motivation for this one is https://github.com/pytorch/pytorch/issues/106789

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -825,8 +825,6 @@ Tensor prod_backward(
   Tensor zero_idx = (input == 0).nonzero();
   if (zero_idx.sym_numel() == 0) {
     return grad * (result / input).conj();
-  } else if (zero_idx.size(0) > 1) {
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:828]: (!at::GradMode::is_enabled() && zero_idx.size(0) > 1)

**Review by albanD — COMMENTED** [review]
Ho ok!
You can update the condition to only be used when double backward is not used then (see condition below)
You can also add a small comment to that effect: when setting up for double backward, we must do the hard work of properly computing the result even though we know it is going to all 0s to ensure the autograd graph is properly created.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (guilhermeleobas) [reply torch/csrc/autograd/FunctionsManual.cpp:828]: Thanks for the tip, @albanD. I've updated the code.

**Review by albanD — COMMENTED** [review]
Thanks!
I would expect that there is some test that used to fail that is now passing?

**Comment by guilhermeleobas** [comment]
@albanD, yes! The following fails on PyTorch main but works on this branch:

```python
import torch

x = torch.tensor([2., 3, 0, 0], requires_grad=True)
y = torch.cumprod(x, dim=0)
gx, = torch.autograd.grad(y.sum(), x, create_graph=True)
gy = torch.autograd.grad(gx.sum(), x)
print(gy)
```

**Comment by albanD** [comment]
Ho I meant there should be a test in CI running this.

**Comment by guilhermeleobas** [comment]
The extra input I added on common_method_invocarions should do it, no?

**Review by albanD — APPROVED** [review]
> The extra input I added on common_method_invocarions should do it, no?

That was my question :p  So we already do have gradgradcheck check for this input and so adding this input without the update in c++ would fail?
If so, then SGTM!

**Comment by guilhermeleobas** [comment]
> > The extra input I added on common_method_invocarions should do it, no?
> 
> That was my question :p So we already do have gradgradcheck check for this input and so adding this input without the update in c++ would fail? If so, then SGTM!

Yep! `python test/test_ops_gradients.py -k test_fn_gradgrad_prod_cpu_float64`

Thanks for the review, @albanD. I'll merge this one.

**Pull request closed without merging**